From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH 14/15] ocfs2: add the ability to switch between heartbeat group modes

 This patch allows the user to switch heartbeat modes by writing the desired
 mode into /sys/o2cb/heartbeat_mode.

 The mode cannot be switched when another is in use. This means that
 <configfs>/cluster/ must be empty.

 fs/ocfs2/cluster/heartbeat.h |    3 +++
 fs/ocfs2/cluster/sys.c       |   18 ++++++++++++++++--
 2 files changed, 19 insertions(+), 2 deletions(-)

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

--- a/fs/ocfs2/cluster/heartbeat.h	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/heartbeat.h	2007-08-27 14:01:22.000000000 -0400
@@ -106,6 +106,9 @@ int o2hb_check_node_heartbeating_from_ca
                                                u8 node_num);
 int o2hb_check_local_node_heartbeating(const char *resource);
 
+const char *o2hb_heartbeat_mode(void);
+int o2hb_set_heartbeat_mode(const char *type, size_t count);
+
 struct o2hb_heartbeat_resource *o2hb_heartbeat_resource_get_by_name(const char * name);
 
 static inline struct o2hb_heartbeat_group *to_o2hb_heartbeat_group(struct config_group *group)
--- a/fs/ocfs2/cluster/sys.c	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/sys.c	2007-08-27 14:01:22.000000000 -0400
@@ -1,5 +1,4 @@
-/* -*- mode: c; c-basic-offset: 8; -*-
- * vim: noexpandtab sw=8 ts=8 sts=0:
+/* -*- mode: c; c-basic-offset: 8; -*- * vim: noexpandtab sw=8 ts=8 sts=0:
  *
  * sys.c
  *
@@ -30,6 +29,7 @@
 #include <linux/sysfs.h>
 
 #include "ocfs2_nodemanager.h"
+#include "heartbeat.h"
 #include "quorum.h"
 #include "masklog.h"
 #include "sys.h"
@@ -75,9 +75,23 @@ static ssize_t o2cb_fence_method_store(c
 static O2CB_ATTR(fence_method, S_IFREG | S_IRUGO | S_IWUSR,
                  o2cb_fence_method_show, o2cb_fence_method_store);
 
+static ssize_t o2cb_heartbeat_mode_show(char *buf)
+{
+	return snprintf(buf, PAGE_SIZE, "%s\n", o2hb_heartbeat_mode());
+}
+
+static ssize_t o2cb_heartbeat_mode_store(const char * buffer, size_t count)
+{
+	return o2hb_set_heartbeat_mode(buffer, count);
+}
+
+static O2CB_ATTR(heartbeat_mode, S_IFREG | S_IRUGO | S_IWUSR,
+                 o2cb_heartbeat_mode_show, o2cb_heartbeat_mode_store);
+
 static struct attribute *o2cb_attrs[] = {
 	&o2cb_attr_interface_revision.attr,
 	&o2cb_attr_fence_method.attr,
+	&o2cb_attr_heartbeat_mode.attr,
 	NULL,
 };
 
